### DISCUSSION TITLE
How JanusGraph balance habse storage

### DISCUSSION THREAD

**Discussion by dinhphu2k1-gif** [JanusGraph/janusgraph#3318] [Q&A]
I use Hbase as storeage backend. I don't know how Janusgraph choose id for a vertex. 
I have a problem that request read and write on regions are differrent. Some regions have a large number of requests read, and some do the opposite. In general, they are not balanced
Can someone give me an idea to rebalance hbase?

**Comment by vtslab** [comment]
Relevant sections in the ref docs are:
- https://docs.janusgraph.org/advanced-topics/partitioning
- cluster.max-partitions       on https://docs.janusgraph.org/configs/configuration-reference
- storage.hbase.region-count
- storage.hbase.regions-per-server

Also note that automatic HBase compaction may "destroy" any planned partitioning. E.g. if you have 20 region servers that each store 1 GB of data after the initial load, HBase will try to compact that to one 20 GB region server if not configured otherwise. HBase can do this for overall cluster efficiency for all tables and applications it serves, but that will not profit your particular JanusGraph app.

**Reply by dinhphu2k1-gif** [reply]
@vtslab 
Is the partition in janusgraph the same as the region in hbase, and the data stored on them is the same?
And does janusgraph hash the values in the vertex to generate the key or does it randomly choose the id as the key to store in the hbase region?

**Reply by vtslab** [reply]
JanusGraph can utilize the entire space of row keys and implemented a centralized id manager that issues blocks of successive id's. cluster.max-partitions is agnostic of the type of storage backend and determines the number of partitions that the id manager envisions in allocating the id blocks. Partitions are randomly assigned to nodes and indeed, for the specific case of the HBase backend a node translates into a region server. 
Note that a physical or virtual machine may host multiple HBase region servers, but for a well designed HBase cluster this should not matter (it does not matter much whether a region server shares resources with a region server of the same application or another application).

**Reply by dinhphu2k1-gif** [reply]
![image](https://user-images.githubusercontent.com/81508954/201568249-490ca363-0d6c-44e7-904e-1b4589790c5f.png)
@vtslab do you know how to decode my data saved in hbase?
Or you know query a record in Hbase by using either vertex id or any properties of this vertex

**Comment by vtslab** [comment]
When I do a google search on "decode janusgraph row key", the following result is ranked highest:
https://gist.github.com/bsnuag/7cce8a99171ec798f7d4e2f1437399d9

You can give it a try.

**Reply by dinhphu2k1-gif** [reply]
@vtslab  Is the edge in the same region as the vertex?
And how i can find vertex in hbase or check that vertex of what region in hbase
